pandas/tests/sparse/test_array.py

```diff
@@ -377,33 +402,54 @@ def test_constructor_float32(self):
             tm.assert_numpy_array_equal(dense, data)
 
     def test_astype(self):
-        res = self.arr.astype('f8')
-        res.sp_values[:3] = 27
-        assert not (self.arr.sp_values[:3] == 27).any()
+        # float -> float
+        arr = SparseArray([None, None, 0, 2])
+        result = arr.astype("Sparse[float32]")
+        expected = SparseArray([None, None, 0, 2], dtype=np.dtype('float32'))
+        tm.assert_sp_array_equal(result, expected)
 
-        msg = "unable to coerce current fill_value nan to int64 dtype"
-        with tm.assert_raises_regex(ValueError, msg):
-            self.arr.astype('i8')
+        dtype = SparseDtype("float64", fill_value=0)
+        result = arr.astype(dtype)
+        expected = SparseArray._simple_new(np.array([0., 2.],
+                                                    dtype=dtype.subtype),
+                                           IntIndex(4, [2, 3]),
+                                           dtype)
+        tm.assert_sp_array_equal(result, expected)
 
-        arr = SparseArray([0, np.nan, 0, 1])
-        with tm.assert_raises_regex(ValueError, msg):
-            arr.astype('i8')
+        dtype = SparseDtype("int64", 0)
+        result = arr.astype(dtype)
+        expected = SparseArray._simple_new(np.array([0, 2], dtype=np.int64),
+                                           IntIndex(4, [2, 3]),
+                                           dtype)
+        tm.assert_sp_array_equal(result, expected)
 
         arr = SparseArray([0, np.nan, 0, 1], fill_value=0)
-        msg = 'Cannot convert non-finite values \\(NA or inf\\) to integer'
-        with tm.assert_raises_regex(ValueError, msg):
-            arr.astype('i8')
+        with tm.assert_raises_regex(ValueError, 'NA'):
+            arr.astype('Sparse[i8]')
+
+    def test_astype_bool(self):
+        a = pd.SparseArray([1, 0, 0, 1], dtype=SparseDtype(int, 0))
+        result = a.astype(bool)
+        expected = SparseArray([True, 0, 0, True],
+                               dtype=SparseDtype(bool, 0))
```

Comment:
This is a "regression" (or at least backwards incompatible change):

```
In [71]: a = pd.SparseArray([1, 0, 0, 1])

In [72]: a
Out[72]: 
[1, 0, 0, 1]
Fill: 0
IntIndex
Indices: array([0, 3], dtype=int32)

In [73]: a.astype(bool)
Out[73]: 
[True, False, False, True]
Fill: False
IntIndex
Indices: array([0, 3], dtype=int32)
```

I personally find it logical to also "astype" the fill value (not changing a completely different value, like swapping the default 0 for ints with default nan for float, but just converting it (if possible) to the other type)

Reply:
This will be swallowed up by the "having a fill_value with a different dtype from sp_values" discussion, so postponing if that's OK.